### DISCUSSION TITLE
Please implement appropriate traits to use `Timestamptz` with `time::OffsetDateTime`

### DISCUSSION THREAD

**Discussion by coriolinus** [diesel-rs/diesel#3241] [Ideas]
I'm working on a project which is using Diesel. We're trying to move away from the `chrono` crate due to its `cargo audit` failures, to `time`. However, I'm currently blocked on that project because Diesel does not fully support it.

Excerpt from `Cargo.toml`:

```toml
[dependencies]
diesel = { version = "1.4", features = ["postgres", "r2d2", "uuidv07", "time", "serde_json"] }
time = { version = "0.3.11", features = ["parsing", "macros"] }
```

The expected outcome is that with features `"time"` and `"postgres"`, diesel supports the `time` crate equivalently to its support for `chrono`. In particular, it should be possible to use an [`OffsetDateTime`](https://docs.rs/time/latest/time/struct.OffsetDateTime.html) with a [`Timestamptz`](https://docs.rs/diesel/latest/diesel/pg/types/sql_types/struct.Timestamptz.html) column.

The actual outcome is several errors of the form:

```
error[E0277]: the trait bound `OffsetDateTime: diesel::Queryable<diesel::sql_types::Timestamptz, Pg>` is not satisfied
    --> src/db/mod.rs:203:14
     |
203  |             .first(&connection)
     |              ^^^^^ the trait `diesel::Queryable<diesel::sql_types::Timestamptz, Pg>` is not implemented for `OffsetDateTime`
     |
     = help: the following other types implement trait `diesel::Queryable<ST, DB>`:
               <(A, B) as diesel::Queryable<(SA, SB), __DB>>
               <(A, B) as diesel::Queryable<diesel::sql_types::Record<(SA, SB)>, Pg>>
               <(A, B, C) as diesel::Queryable<(SA, SB, SC), __DB>>
               <(A, B, C) as diesel::Queryable<diesel::sql_types::Record<(SA, SB, SC)>, Pg>>
               <(A, B, C, D) as diesel::Queryable<(SA, SB, SC, SD), __DB>>
               <(A, B, C, D) as diesel::Queryable<diesel::sql_types::Record<(SA, SB, SC, SD)>, Pg>>
               <(A, B, C, D, E) as diesel::Queryable<(SA, SB, SC, SD, SE), __DB>>
               <(A, B, C, D, E) as diesel::Queryable<diesel::sql_types::Record<(SA, SB, SC, SD, SE)>, Pg>>
             and 96 others
     = note: required because of the requirements on the impl of `diesel::Queryable<(diesel::sql_types::Uuid, diesel::sql_types::Jsonb, diesel::sql_types::Integer, diesel::sql_types::Timestamptz, diesel::sql_types::Timestamptz), Pg>` for `(models::legal_entity_id::LegalEntityId, serde_json::Value, i32, OffsetDateTime, OffsetDateTime)`
     = note: 1 redundant requirement hidden
     = note: required because of the requirements on the impl of `diesel::Queryable<(diesel::sql_types::Uuid, diesel::sql_types::Jsonb, diesel::sql_types::Integer, diesel::sql_types::Timestamptz, diesel::sql_types::Timestamptz), Pg>` for `models::legal_entity::LegalEntity`
     = note: required because of the requirements on the impl of `LoadQuery<PooledConnection<ConnectionManager<PgConnection>>, models::legal_entity::LegalEntity>` for `diesel::query_builder::SelectStatement<schema::legal_entities::table, query_builder::select_clause::DefaultSelectClause, query_builder::distinct_clause::NoDistinctClause, query_builder::where_clause::WhereClause<diesel::expression::operators::Eq<schema::legal_entities::columns::id, diesel::expression::bound::Bound<diesel::sql_types::Uuid, models::legal_entity_id::LegalEntityId>>>, query_builder::order_clause::NoOrderClause, query_builder::limit_clause::LimitClause<diesel::expression::bound::Bound<BigInt, i64>>>`
note: required by a bound in `first`
    --> /home/coriolinus/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-1.4.8/src/query_dsl/mod.rs:1343:22
     |
1343 |         Limit<Self>: LoadQuery<Conn, U>,
     |                      ^^^^^^^^^^^^^^^^^^ required by this bound in `first`
```

The expected list of trait implementations required:

- [`impl ToSql<TimestampTz, Pg> for OffsetDateTime`](https://docs.rs/diesel/latest/diesel/serialize/trait.ToSql.html)
- [`impl FromSql<TimestampTz, Pg> for OffsetDateTime`](https://docs.rs/diesel/latest/diesel/deserialize/trait.FromSql.html)

**Comment by weiznich** [comment]
Diesel 1.4.x does only support version 0.1.x of the time crate. Support for that was completely removed for the upcoming 2.0 release, due to the fact that time 0.1 is long deprecated and the newer 0.3 releases do not have much in common with the old crate beside the name. 
We are open to adding support for time 0.3 if someone submits a PR containing code, tests and documentation. Otherwise I think that might not be something that's high up on our priority list currently.

**Reply by coriolinus** [reply]
I understand. 

Is there a resource for how to get started on that PR? I'm not sure whether I can convince my boss that my time is best spent writing a Diesel feature right now, but if I can, it would be great not to jump in blindly.

**Reply by weiznich** [reply]
It shouldn't be that hard to provide such an implementation. I think the following parts need to be done

1. Add `time` as optional dependency here: https://github.com/diesel-rs/diesel/blob/44c7f3f3273fd6492340c2168ab832f8a742acef/diesel/Cargo.toml#L15
2. Provide a version of these `chrono` specific impls for `time`: https://github.com/diesel-rs/diesel/blob/44c7f3f3273fd6492340c2168ab832f8a742acef/diesel/src/pg/types/date_and_time/chrono.rs#L1
This essentially can just copy and adjust the chrono module + all tests there. The only part there that requires some real implementation work is to figure out how to convert from/to the relevant time types in the relevant `FromSql`/`ToSql` impls.
4. Similarly for the sqlite backend as well: 
https://github.com/diesel-rs/diesel/blob/44c7f3f3273fd6492340c2168ab832f8a742acef/diesel/src/sqlite/types/date_and_time/chrono.rs#L1
6. For the mysql backend as well here: https://github.com/diesel-rs/diesel/blob/44c7f3f3273fd6492340c2168ab832f8a742acef/diesel/src/mysql/types/date_and_time.rs#L139-L259 (that one might require slight refactoring such that it follows the same module layout as the postgres and sqlite impls)
7. Enable the `time` feature on the CI here https://github.com/diesel-rs/diesel/blob/44c7f3f3273fd6492340c2168ab832f8a742acef/.github/workflows/ci.yml#L1
8. Add a changelog entry here: https://github.com/diesel-rs/diesel/blob/44c7f3f3273fd6492340c2168ab832f8a742acef/CHANGELOG.md?plain=1#L13

If you hit any problem while implementing this feature we are happy to answer questions or provide specific help.

**Reply by coriolinus** [reply]
I'm having trouble with step 3: there's a WIP implementation [here](https://github.com/coriolinus/diesel/blob/d143170d7087d3d461731ff2e1942cd5378285ef/diesel/src/sqlite/types/date_and_time/time.rs), but the tests do not pass, and it's really not clear why. Current status:

```sh
$ cargo test --features time --features sqlite -- sqlite::types::date_and_time::time
Fri 22 Jul 2022 05:02:42 PM CEST
    Blocking waiting for file lock on build directory
   Compiling diesel v2.0.0-rc.0 (/home/coriolinus/projects/diesel-rs/diesel/diesel)
    Finished test [unoptimized + debuginfo] target(s) in 10.21s
     Running unittests (/home/coriolinus/projects/diesel-rs/diesel/target/debug/deps/diesel-48b6a9d23676a63f)

running 14 tests
test sqlite::types::date_and_time::time::tests::times_of_day_encode_correctly ... FAILED
test sqlite::types::date_and_time::time::tests::dates_decode_correctly ... ok
test sqlite::types::date_and_time::time::tests::datetimes_encode_correctly ... FAILED
test sqlite::types::date_and_time::time::tests::dates_encode_correctly ... ok
test sqlite::types::date_and_time::time::tests::datetimes_decode_correctly ... ok
test sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly ... FAILED
test sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly_with_timezone ... ok
test sqlite::types::date_and_time::time::tests::times_relative_to_now_encode_correctly ... ok
test sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly_with_utc_timezone ... FAILED
test sqlite::types::date_and_time::time::tests::insert_timestamptz_into_table_as_text ... ok
test sqlite::types::date_and_time::time::tests::can_query_timestamptz_column_with_between ... ok
test sqlite::types::date_and_time::time::tests::times_of_day_decode_correctly ... ok
test sqlite::types::date_and_time::time::tests::unix_epoch_decodes_correctly_with_utc_timezone_in_all_possible_formats ... ok
test sqlite::types::date_and_time::time::tests::unix_epoch_decodes_correctly_in_all_possible_formats ... ok

failures:

---- sqlite::types::date_and_time::time::tests::times_of_day_encode_correctly stdout ----
SELECT (time(?) = ?) -- binds: ["00:00:00.000000", Time { hour: 0, minute: 0, second: 0, nanosecond: 0 }]
thread 'sqlite::types::date_and_time::time::tests::times_of_day_encode_correctly' panicked at 'assertion failed: `(left == right)`
  left: `false`,
 right: `true`', diesel/src/sqlite/types/date_and_time/time.rs:319:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- sqlite::types::date_and_time::time::tests::datetimes_encode_correctly stdout ----
thread 'sqlite::types::date_and_time::time::tests::datetimes_encode_correctly' panicked at 'assertion failed: query.get_result::<bool>(connection).unwrap()', diesel/src/sqlite/types/date_and_time/time.rs:464:9

---- sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly stdout ----
thread 'sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly' panicked at 'assertion failed: `(left == right)`
  left: `Ok(true)`,
 right: `Ok(false)`', diesel/src/sqlite/types/date_and_time/time.rs:242:9

---- sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly_with_utc_timezone stdout ----
thread 'sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly_with_utc_timezone' panicked at 'assertion failed: query.get_result::<bool>(connection).unwrap()', diesel/src/sqlite/types/date_and_time/time.rs:599:9


failures:
    sqlite::types::date_and_time::time::tests::datetimes_encode_correctly
    sqlite::types::date_and_time::time::tests::times_of_day_encode_correctly
    sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly
    sqlite::types::date_and_time::time::tests::unix_epoch_encodes_correctly_with_utc_timezone

test result: FAILED. 10 passed; 4 failed; 0 ignored; 0 measured; 24 filtered out; finished in 0.01s

error: test failed, to rerun pass '--lib'
```

It's very hard to tell why the failing tests are in fact failing. `debug_query` is not proving very useful here; it produces output like 

```
[diesel/src/sqlite/types/date_and_time/time.rs:318] debug_query(&query) = Query {
    sql: "SELECT (time(?) = ?)",
    binds: [
        "00:00:00.000000",
        Time {
            hour: 0,
            minute: 0,
            second: 0,
            nanosecond: 0,
        },
    ],
}
```

which _looks_ like it should pass.

Are there other test fixtures or resources I could use to help debug this and get the tests passing?

**Reply by weiznich** [reply]
Thanks for starting to implement this :heart:

Some short ideas for the test failures before leaving for weekend: I would likely start with checking if the corresponding [ToSql](https://github.com/coriolinus/diesel/blob/d143170d7087d3d461731ff2e1942cd5378285ef/diesel/src/sqlite/types/date_and_time/time.rs#L120) impl generates the expected value by adding a `dbg!` call there. If that's the case I would check if the sql query with the relevant bind values inserted returns true.

**Reply by coriolinus** [reply]
I'm also going to put this down for the weekend soon, but I'm just getting more confused.

Test output with `dbg` in the `ToSql` impl:

```
---- sqlite::types::date_and_time::time::tests::times_of_day_encode_correctly stdout ----
[diesel/src/sqlite/types/date_and_time/time.rs:119] self.format(ENCODE_TIME_FORMAT).map_err(|err| err.to_string()) = Ok(
    "00:00:00.000000",
)
thread 'sqlite::types::date_and_time::time::tests::times_of_day_encode_correctly' panicked at 'assertion failed: `(left == right)`
  left: `false`,
 right: `true`', diesel/src/sqlite/types/date_and_time/time.rs:310:9
```

This output looks identical to what's expected. The test failure is generated by the query with relevant bind values returning false: https://github.com/diesel-rs/diesel/blob/b2d7b7b30dbe7aef2fda6d3800e100d2bc3b32f8/diesel/src/sqlite/types/date_and_time/time.rs#L309-L310

I wonder if it's a type issue? The way the `time` function in that expression works is very mysterious to me. https://github.com/diesel-rs/diesel/blob/b2d7b7b30dbe7aef2fda6d3800e100d2bc3b32f8/diesel/src/sqlite/types/date_and_time/time.rs#L226

**Comment by coriolinus** [comment]
Resolved by https://github.com/diesel-rs/diesel/pull/3246.
